osquery/tables/system/windows/ntfs_acl_permissions.cpp

```diff
@@ -0,0 +1,184 @@
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <AccCtrl.h>
+#include <Aclapi.h>
+#include <map>
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include <windows.h>
+
+#include <boost/algorithm/string/join.hpp>
+#include <boost/filesystem.hpp>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+namespace alg = boost::algorithm;
+namespace fs = boost::filesystem;
+
+namespace osquery {
+namespace tables {
+
+// map to get access mode string
+static std::unordered_map<ACCESS_MODE, std::string> accessModeToStr = {
+    {NOT_USED_ACCESS, "Not Used"},
+    {GRANT_ACCESS, "Grant"},
+    {SET_ACCESS, "Set"},
+    {DENY_ACCESS, "Deny"},
+    {REVOKE_ACCESS, "Revoke"},
+    {SET_AUDIT_SUCCESS, "Set Audit Success"},
+    {SET_AUDIT_FAILURE, "Set Audit Failure"}};
+
+// map to build access string
+static std::map<unsigned long, std::string> permVals = {
+    {DELETE, "Delete"},
+    {READ_CONTROL, "Read Control"},
+    {WRITE_DAC, "Write DAC"},
+    {WRITE_OWNER, "Write Owner"},
+    {SYNCHRONIZE, "Synchronize"},
+    {STANDARD_RIGHTS_REQUIRED, "Std Rights Required"},
+    {STANDARD_RIGHTS_ALL, "Std Rights All"},
+    {SPECIFIC_RIGHTS_ALL, "Specific Rights All"},
+    {ACCESS_SYSTEM_SECURITY, "Access System Security"},
+    {MAXIMUM_ALLOWED, "Maximum Allowed"},
+    {GENERIC_READ, "Generic Read"},
+    {GENERIC_WRITE, "Generic Write"},
+    {GENERIC_EXECUTE, "Generic Execute"},
+    {GENERIC_ALL, "Generic All"}};
+
+// map to get inheritance string
+static std::unordered_map<unsigned long, std::string> inheritanceToStr = {
+    {CONTAINER_INHERIT_ACE, "Container Inherit Ace"},
+    {INHERIT_NO_PROPAGATE, "Inherit No Propagate"},
+    {INHERIT_ONLY, "Inherit Only"},
+    {OBJECT_INHERIT_ACE, "Object Inherit Ace"},
+    {SUB_CONTAINERS_AND_OBJECTS_INHERIT, "Sub containers and Objects Inherit"},
+    {NO_INHERITANCE, "No Inheritence"}};
+
+// helper function to build access string from permission bit mask
+std::string accessPermsToStr(const unsigned long pmask) {
+  std::vector<std::string> permList;
+
+  for (auto const& perm : permVals) {
+    if ((pmask & perm.first) != 0) {
+      permList.push_back(perm.second);
+    }
+  }
+
+  return alg::join(permList, ",");
+}
+
+// helper function to get account/group name from trustee
+std::string trusteeToStr(const TRUSTEE& trustee) {
+  // Max username length for Windows
+  int r;
+  const unsigned long kMaxBuffSize = 256;
+  char name[kMaxBuffSize];
+  char domain[kMaxBuffSize];
+  unsigned long size = kMaxBuffSize;
+  SID_NAME_USE accountType;
+
+  switch (trustee.TrusteeForm) {
+  case TRUSTEE_IS_SID: {
+    // get the name from the SID
+    PSID psid = trustee.ptstrName;
+    r = LookupAccountSid(
+        nullptr, psid, name, &size, domain, &size, &accountType);
```

Comment:
As far as I can see [here](https://msdn.microsoft.com/en-us/library/windows/desktop/aa379166(v=vs.85).aspx) 4th and 6th arguments are input-output. Function reads both of them and write final length of `name` and `domain` to it. This is a place of potential mistakes.

For instance: Function reads the length of name, writes something to `name` with length 4 and changes `size` to 4. Then function reads the length of `domain`, writes smth to `domain` with length 10 and fails because the maximum available space for `domain` is 4.
It is just an example, I don't know how exactly `LookupAccountSid` works. But may be it will explain my concern.
And because of that I think it is a bad idea to put the same pointer to it.

Reply:
Good catch! In theory, this should never happen because both cases should never trigger in the function. But I'll create a separate size for each case for good measure.